Test failures in test_trsm

Test failures on travis-ci https://travis-ci.org/MacPython/scipy-wheels/builds/283669544:
```
=================================== FAILURES ===================================
__________________________________ test_trsm ___________________________________

    def test_trsm():
        seed(1234)
        for ind, dtype in enumerate(DTYPES):
            func, = get_blas_funcs(('trsm',), dtype=dtype)
    
            # Test protection against size mismatches
            A = rand(4, 5).astype(dtype)
            B = rand(4, 4).astype(dtype)
            alpha = dtype(1)
            assert_raises(Exception, func, alpha, A, B)
            assert_raises(Exception, func, alpha, A.T, B)
    
            n = 8
            m = 7
            alpha = dtype(-2.5)
            A = (rand(m, m) if ind < 2 else rand(m, m) + rand(m, m)*1j) + eye(m)
            A = A.astype(dtype)
            Au = triu(A)
            Al = tril(A)
            B1 = rand(m, n).astype(dtype)
            B2 = rand(n, m).astype(dtype)
    
            x1 = func(alpha=alpha, a=A, b=B1)
            assert_equal(B1.shape, x1.shape)
            x2 = solve(Au, alpha*B1)
            assert_array_almost_equal(x1, x2)
    
            x1 = func(alpha=alpha, a=A, b=B1, trans_a=1)
            x2 = solve(Au.T, alpha*B1)
            assert_array_almost_equal(x1, x2)
    
            x1 = func(alpha=alpha, a=A, b=B1, trans_a=2)
            x2 = solve(Au.conj().T, alpha*B1)
            assert_array_almost_equal(x1, x2)
    
            x1 = func(alpha=alpha, a=A, b=B1, diag=1)
            Au[arange(m), arange(m)] = dtype(1)
            x2 = solve(Au, alpha*B1)
            assert_array_almost_equal(x1, x2)
    
            x1 = func(alpha=alpha, a=A, b=B2, diag=1, side=1)
            x2 = solve(Au.conj().T, alpha*B2.conj().T)
>           assert_array_almost_equal(x1, x2.conj().T)

A          = array([[ 1.48305285+0.61744374j,  0.12533161+0.13226691j,
         0.96445417+...j,  0.12706631+0.06753922j,
         1.41398394+0.31758606j]], dtype=complex64)
Al         = array([[ 1.48305285+0.61744374j,  0.00000000+0.j        ,
         0.00000000+...j,  0.12706631+0.06753922j,
         1.41398394+0.31758606j]], dtype=complex64)
Au         = array([[ 1.00000000+0.j        ,  0.12533161+0.13226691j,
         0.96445417+...j        ,  0.00000000+0.j        ,  1.00000000+0.j        ]], dtype=complex64)
B          = array([[ 0.26123163+0.j,  0.93845558+0.j,  0.93864244+0.j,  0.74504453+0.j],
 ...4299+0.j,  0.83358687+0.j,  0.36994851+0.j,  0.93655699+0.j]], dtype=complex64)
B1         = array([[ 0.69686300+0.j,  0.64026541+0.j,  0.63262331+0.j,  0.82875067+0.j,
  ...6132+0.j,  0.83229774+0.j,  0.34565136+0.j,  0.23997119+0.j]], dtype=complex64)
B2         = array([[ 0.53578758+0.j,  0.27146694+0.j,  0.03443087+0.j,  0.62708557+0.j,
  ...,
         0.02454439+0.j,  0.25736493+0.j,  0.41011602+0.j]], dtype=complex64)
alpha      = (-2.5+0j)
dtype      = <type 'numpy.complex64'>
func       = <fortran object>
ind        = 2
m          = 7
n          = 8
x1         = array([[-1.33946896+0.j        , -0.51078957+0.17716742j,
         1.64277339+...j, -1.55378079-0.60774016j,
        -1.30365491+1.4840529j ]], dtype=complex64)
x2         = array([[-1.33946872 -1.19209290e-07j, -2.47087431 -3.57627869e-07j,
        -1... -1.28172421 +9.26676750e-01j, -1.30365443 -1.48405361e+00j]], dtype=complex64)

/venv/local/lib/python2.7/site-packages/scipy/linalg/tests/test_blas.py:1072: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/venv/local/lib/python2.7/site-packages/numpy/testing/utils.py:811: in assert_array_almost_equal
    header=('Arrays are not almost equal to %d decimals' % decimal))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

comparison = <function compare at 0x7fe9fec68230>
x = array([[-1.33946896+0.j        , -0.51078957+0.17716742j,
         1.64277339+...j, -1.55378079-0.60774016j,
        -1.30365491+1.4840529j ]], dtype=complex64)
y = array([[-1.33946872 +1.19209290e-07j, -0.51078969 +1.77167475e-01j,
         1...8163 -6.07739985e-01j,
        -1.30365443 +1.48405361e+00j]], dtype=complex64)
err_msg = '', verbose = True
header = 'Arrays are not almost equal to 6 decimals'

    def assert_array_compare(comparison, x, y, err_msg='', verbose=True,
                             header=''):
        from numpy.core import array, isnan, isinf, any, all, inf
        x = array(x, copy=False, subok=True)
        y = array(y, copy=False, subok=True)
    
        def isnumber(x):
            return x.dtype.char in '?bhilqpBHILQPefdgFDG'
    
        def chk_same_position(x_id, y_id, hasval='nan'):
            """Handling nan/inf: check that x and y have the nan/inf at the same
            locations."""
            try:
                assert_array_equal(x_id, y_id)
            except AssertionError:
                msg = build_err_msg([x, y],
                                    err_msg + '\nx and y %s location mismatch:' \
                                    % (hasval), verbose=verbose, header=header,
                                    names=('x', 'y'))
                raise AssertionError(msg)
    
        try:
            cond = (x.shape==() or y.shape==()) or x.shape == y.shape
            if not cond:
                msg = build_err_msg([x, y],
                                    err_msg
                                    + '\n(shapes %s, %s mismatch)' % (x.shape,
                                                                      y.shape),
                                    verbose=verbose, header=header,
                                    names=('x', 'y'))
                if not cond :
                    raise AssertionError(msg)
    
            if isnumber(x) and isnumber(y):
                x_isnan, y_isnan = isnan(x), isnan(y)
                x_isinf, y_isinf = isinf(x), isinf(y)
    
                # Validate that the special values are in the same place
                if any(x_isnan) or any(y_isnan):
                    chk_same_position(x_isnan, y_isnan, hasval='nan')
                if any(x_isinf) or any(y_isinf):
                    # Check +inf and -inf separately, since they are different
                    chk_same_position(x == +inf, y == +inf, hasval='+inf')
                    chk_same_position(x == -inf, y == -inf, hasval='-inf')
    
                # Combine all the special values
                x_id, y_id = x_isnan, y_isnan
                x_id |= x_isinf
                y_id |= y_isinf
    
                # Only do the comparison if actual values are left
                if all(x_id):
                    return
    
                if any(x_id):
                    val = comparison(x[~x_id], y[~y_id])
                else:
                    val = comparison(x, y)
            else:
                val = comparison(x, y)
    
            if isinstance(val, bool):
                cond = val
                reduced = [0]
            else:
                reduced = val.ravel()
                cond = reduced.all()
                reduced = reduced.tolist()
            if not cond:
                match = 100-100.0*reduced.count(1)/len(reduced)
                msg = build_err_msg([x, y],
                                    err_msg
                                    + '\n(mismatch %s%%)' % (match,),
                                    verbose=verbose, header=header,
                                    names=('x', 'y'))
                if not cond :
>                   raise AssertionError(msg)
E                   AssertionError: 
E                   Arrays are not almost equal to 6 decimals
E                   
E                   (mismatch 1.78571428571%)
E                    x: array([[-1.33946896+0.j        , -0.51078957+0.17716742j,
E                            1.64277339+1.41887403j, -0.00436068-1.32258463j,
E                           -0.07933950-0.35224861j, -1.51080155+0.64535797j,...
E                    y: array([[-1.33946872 +1.19209290e-07j, -0.51078969 +1.77167475e-01j,
E                            1.64277363 +1.41887379e+00j, -0.00436084 -1.32258475e+00j,
E                           -0.07933977 -3.52248728e-01j, -1.51080155 +6.45358443e-01j,...

all        = <function all at 0x7fea31b6b6e0>
any        = <function any at 0x7fea31b6b668>
array      = <built-in function array>
chk_same_position = <function chk_same_position at 0x7fe9fec682a8>
comparison = <function compare at 0x7fe9fec68230>
cond       = False
err_msg    = ''
header     = 'Arrays are not almost equal to 6 decimals'
inf        = inf
isinf      = <ufunc 'isinf'>
isnan      = <ufunc 'isnan'>
isnumber   = <function isnumber at 0x7fe9fec68758>
match      = 1.7857142857142918
msg        = '\nArrays are not almost equal to 6 decimals\n\n(mismatch 1.78571428571%)\n x: array([[-1.33946896+0.j        , -0.510...1.41887379e+00j, -0.00436084 -1.32258475e+00j,\n        -0.07933977 -3.52248728e-01j, -1.51080155 +6.45358443e-01j,...'
reduced    = [True, True, True, True, True, True, ...]
val        = array([[ True,  True,  True,  True,  True,  True,  True],
       [ True,  True...  True],
       [ True,  True,  True,  True,  True,  True,  True]], dtype=bool)
verbose    = True
x          = array([[-1.33946896+0.j        , -0.51078957+0.17716742j,
         1.64277339+...j, -1.55378079-0.60774016j,
        -1.30365491+1.4840529j ]], dtype=complex64)
x_id       = array([[False, False, False, False, False, False, False],
       [False, False... False],
       [False, False, False, False, False, False, False]], dtype=bool)
x_isinf    = array([[False, False, False, False, False, False, False],
       [False, False... False],
       [False, False, False, False, False, False, False]], dtype=bool)
x_isnan    = array([[False, False, False, False, False, False, False],
       [False, False... False],
       [False, False, False, False, False, False, False]], dtype=bool)
y          = array([[-1.33946872 +1.19209290e-07j, -0.51078969 +1.77167475e-01j,
         1...8163 -6.07739985e-01j,
        -1.30365443 +1.48405361e+00j]], dtype=complex64)
y_id       = array([[False, False, False, False, False, False, False],
       [False, False... False],
       [False, False, False, False, False, False, False]], dtype=bool)
y_isinf    = array([[False, False, False, False, False, False, False],
       [False, False... False],
       [False, False, False, False, False, False, False]], dtype=bool)
y_isnan    = array([[False, False, False, False, False, False, False],
       [False, False... False],
       [False, False, False, False, False, False, False]], dtype=bool)

/venv/local/lib/python2.7/site-packages/numpy/testing/utils.py:644: AssertionError
 1 failed, 12544 passed, 1591 skipped, 97 xfailed, 9 xpassed in 697.34 seconds =
```

These failures probably are sporadic. Maybe the default tolerance for `assert_almost_equal` is too small for single precision?

Seems to have been added in gh-7843

--- a/scipy/linalg/tests/test_blas.py
+++ b/scipy/linalg/tests/test_blas.py
@@ -1031,6 +1031,7 @@ class TestTRMM(object):
 def test_trsm():
     seed(1234)
     for ind, dtype in enumerate(DTYPES):
+        tol = np.finfo(dtype).eps*1000
         func, = get_blas_funcs(('trsm',), dtype=dtype)
 
         # Test protection against size mismatches
@@ -1053,27 +1054,26 @@ def test_trsm():
         x1 = func(alpha=alpha, a=A, b=B1)
         assert_equal(B1.shape, x1.shape)
         x2 = solve(Au, alpha*B1)
-        assert_array_almost_equal(x1, x2)
+        assert_allclose(x1, x2, atol=tol)
 
         x1 = func(alpha=alpha, a=A, b=B1, trans_a=1)
         x2 = solve(Au.T, alpha*B1)
-        assert_array_almost_equal(x1, x2)
+        assert_allclose(x1, x2, atol=tol)
 
         x1 = func(alpha=alpha, a=A, b=B1, trans_a=2)
         x2 = solve(Au.conj().T, alpha*B1)
-        assert_array_almost_equal(x1, x2)
+        assert_allclose(x1, x2, atol=tol)
 
         x1 = func(alpha=alpha, a=A, b=B1, diag=1)
         Au[arange(m), arange(m)] = dtype(1)
         x2 = solve(Au, alpha*B1)
-        assert_array_almost_equal(x1, x2)
+        assert_allclose(x1, x2, atol=tol)
 
         x1 = func(alpha=alpha, a=A, b=B2, diag=1, side=1)
-        x2 = solve_triangular(Au.conj().T, alpha*B2.conj().T,
-                              unit_diagonal=1, lower=1)
-        assert_array_almost_equal(x1, x2.conj().T, decimal=5)
+        x2 = solve(Au.conj().T, alpha*B2.conj().T)
+        assert_allclose(x1, x2.conj().T, atol=tol)
 
         x1 = func(alpha=alpha, a=A, b=B2, diag=1, side=1, lower=1)
         Al[arange(m), arange(m)] = dtype(1)
-        x2 = solve_triangular(Al.conj().T, alpha*B2.conj().T, unit_diagonal=1)
-        assert_array_almost_equal(x1, x2.conj().T, decimal=5)
+        x2 = solve(Al.conj().T, alpha*B2.conj().T)
+        assert_allclose(x1, x2.conj().T, atol=tol)

--- a/scipy/linalg/tests/test_blas.py
+++ b/scipy/linalg/tests/test_blas.py
@@ -23,7 +23,8 @@ from numpy import float32, float64, complex64, complex128, arange, triu, \
                   nonzero
 
 from numpy.random import rand, seed
-from scipy.linalg import _fblas as fblas, get_blas_funcs, toeplitz, solve
+from scipy.linalg import _fblas as fblas, get_blas_funcs, toeplitz, solve, \
+                                          solve_triangular
 
 try:
     from scipy.linalg import _cblas as cblas
@@ -1068,10 +1069,11 @@ def test_trsm():
         assert_array_almost_equal(x1, x2)
 
         x1 = func(alpha=alpha, a=A, b=B2, diag=1, side=1)
-        x2 = solve(Au.conj().T, alpha*B2.conj().T)
-        assert_array_almost_equal(x1, x2.conj().T)
+        x2 = solve_triangular(Au.conj().T, alpha*B2.conj().T,
+                              unit_diagonal=1, lower=1)
+        assert_array_almost_equal(x1, x2.conj().T, decimal=5)
 
         x1 = func(alpha=alpha, a=A, b=B2, diag=1, side=1, lower=1)
         Al[arange(m), arange(m)] = dtype(1)
-        x2 = solve(Al.conj().T, alpha*B2.conj().T)
-        assert_array_almost_equal(x1, x2.conj().T)
+        x2 = solve_triangular(Al.conj().T, alpha*B2.conj().T, unit_diagonal=1)
+        assert_array_almost_equal(x1, x2.conj().T, decimal=5)

